vespalib/src/vespa/vespalib/util/ref_counted.h

```diff
@@ -0,0 +1,117 @@
+// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#pragma once
+
+#include <atomic>
+#include <concepts>
+#include <utility>
+
+// This file contains code that implements intrusive reference
+// counting with smart handles (ref_counted<T> points to T, T inherits
+// enable_ref_counted).
+
+// Functions with names starting with 'internal' are not intended for
+// direct use, but are available for completeness (enables incremental
+// re-write of code from bald to smart pointers).
+
+namespace vespalib {
+
+// This is the actual reference count. It cannot be moved or copied,
+// but it can be changed even when const. Classes that want to be
+// counted need to inherit from this. The destructor is virtual to
+// make sure the right one is called. This also enables the subref
+// function to directly delete the shared object.
+class enable_ref_counted
+{
+    static constexpr uint32_t MAGIC = 0xcc56a933;
+private:
+    uint32_t _guard;
+    mutable std::atomic<int32_t> _refs;
+protected:
+    enable_ref_counted() noexcept : _guard(MAGIC), _refs(1) {}
+public:
+    virtual ~enable_ref_counted() noexcept;
+    enable_ref_counted(enable_ref_counted &&) = delete;
+    enable_ref_counted(const enable_ref_counted &) = delete;
+    enable_ref_counted &operator=(enable_ref_counted &&) = delete;
+    enable_ref_counted &operator=(const enable_ref_counted &) = delete;
+    void internal_addref() const noexcept;
+    void internal_subref() const noexcept;
+    int32_t count_refs() const noexcept;
+};
+
+// This is the handle to a shared object. The handle itself is not
+// thread safe.
+template <typename T>
+class ref_counted
+{
+    // carefully placed here to give the best error messages
+    static_assert(std::derived_from<T,enable_ref_counted>);
+    template <typename X> friend class ref_counted;
+private:
+    T *_ptr;
+    ref_counted(T *ptr) noexcept : _ptr(ptr) {}
+    void maybe_subref() noexcept {
+        if (_ptr) {
+            _ptr->internal_subref();
+        }
+    }
+    static T *maybe_addref(T *ptr) noexcept {
+        if (ptr) {
+            ptr->internal_addref();
+        }
+        return ptr;
+    }
+public:
+    ref_counted() noexcept : _ptr(nullptr) {}
+    ref_counted(ref_counted &&rhs) noexcept : _ptr(std::exchange(rhs._ptr, nullptr)) {}
+    ref_counted(const ref_counted &rhs) noexcept : _ptr(maybe_addref(rhs._ptr)) {}
+    ref_counted &operator=(ref_counted &&rhs) noexcept {
+        maybe_subref();
+        _ptr = std::exchange(rhs._ptr, nullptr);
+        return *this;
+    }
+    ref_counted &operator=(const ref_counted &rhs) noexcept {
+        maybe_subref();
+        _ptr = maybe_addref(rhs._ptr);
+        return *this;
+    }
+    template <typename X> ref_counted(ref_counted<X> &&rhs) noexcept : _ptr(std::exchange(rhs._ptr, nullptr)) {}
```

Comment:
Could be pretty neat to have a concept requirement here that `X*` is convertible to `T*` for added compiler helpfulness 🦄 But can be done as a followup, if we want it

Reply:
Disregard this. SFINAE-removal makes things strictly worse when adding a concept requirement here...

Reply:
The concept requirement makes the error message worse since it removes function form the overload set and you end up with an error saying ref_counted<Foo> cannot be converted to ref_counted<Bar> instead of an error saying that Foo* cannot be converted to Bar*.